### ISSUE TITLE
installation tarball can/should be cached to avoid re-downloads

### ISSUE THREAD

**Issue by vn971** [NixOS/nix#540]
The current installation script https://nixos.org/nix/install  does not cache the downloadable ~70Mb installation tarball. This is bad in all cases where installation fails for some reason. Examples:
- a network failure may occur. It may be preferable to resume the download: https://github.com/NixOS/nix/issues/379
- `sudo` command may not be available on the target system, practically failing the installation. This is default for Debian 7-8. https://github.com/NixOS/nix/issues/539
- the script may be run as `root` accidentally. This is not recommended and a user may want to install Nix in userland. He will be forced to re-download the tarball again.
- if you accidentally run the script as `root` first, the `/nix` directory would be owned by `root` and installation will fail again (asking you to `chown`).

The proposed alternative is to cache the tarball. For example, the script could download the tarball first and ask the user whether to delete the tarball on exit. This will save both time and brandwidth for the user.

**Comment by wmertens** [comment]
Ideally it should also use a checksum if available on the system. Perl is available on most systems and has sha256 in core: `perl -e 'use Digest::SHA 'sha256_hex'; print sha256_hex(<tarball.tgz>)'`.

**Comment by vn971** [comment]
@wmertens You mean to avoid out-of-date tarballs?
Basically, tarball caching and verification seem to be different jobs.

**Comment by wmertens** [comment]
No I mean to verify that the tarball is the one that was expected.

**Comment by Anton-Latukha** [comment]
I step up with critique of this issue:
1. As a DevOps I never seen installation oneliner (as https://nixos.org/nix/install) that cache tarballs.

Look at Docker, Salt, Ansible... oneliners. There is no caching. If you find one share with me.

It is not the thing that shell script should do or can do great.

2. Oneliner with caching introduces deliberate sideeffect.
And it must have less side-effects.
I can argue with issue on the contratry: if oneliner failed - it should pull/roll back as nothing happened to the system, and oneliner was not ran.

One of main points and uniqueness of Nix is that it is fully transactional.

3. Caching introduces security concerns.

If do caching, now Nix oneliner has concerns:
3. a. Tarball should be in constant place. So script can use it (or do recurcive search in /tmp). And then this can be exploited, by replacing tarball with malicious one. Than, script must do hash check.
It does now

**Comment by vn971** [comment]
// SHA2 has no second pre-image attacks. Even MD5 has no second pre-image attacks. It's different attacks that are possible, but not second pre-image. https://en.wikipedia.org/wiki/Preimage_attack

Anyway, I personally don't use NixOS anymore, and I don't have the energy to really participate in the issue. Feel free to hijack.

**Comment by wmertens** [comment]
The oneliner downloads a script that then downloads the tarball. It can
provide caching.

**Comment by Anton-Latukha** [comment]
@vn971

Yes, you are right about SHA256. Collision resistance is not the same as preimage attack. As I understand, it can be the first or second one here, as input and function are known.
In papers claimed they can go as far as 52 out of 64 rounds for SHA-256 to break it. Especially since quantum computers become our reality, SHA-256 probably go away as recommended by security in nearest future.

Point of SHA256 is supplementary one, all other points are still legitimate and more important by my view.

**Comment by Anton-Latukha** [comment]
@vn971

I wrote message not because I want to argue. I saw a place to raise does caching needed.
While I find when Nix installation is by itself not in perfect state. It assumes noncalsical modes to install (install as normal user) while sudo with root escalation should be present. List of preassumed conditions are not true for many Linux distributions in clean state (Arch, Alpine, and I see should hit many others). Building from source also requires reverse-engineering on this systems.

And that is where I try to contribute today (working on scripts locally today). While I check bugreports, that I can close on pull request.

**Comment by copumpkin** [comment]
If you're concerned about the security of SHA256, we use it all over the place, so I'd open a separate issue. I personally think it's a tad premature to advocate moving away from it given the state of the world and pain involved though.

**Comment by Anton-Latukha** [comment]
@wmertens 

You can check that now oneliner script does downloads whole tarball.

Under 'oneliner' I mean `https://nixos.org/nix/install` script itself.

**Comment by Anton-Latukha** [comment]
@copumpkin 

As I said, SHA256 is a minor topic at least for me.
I also really see that many other points are important for now.

I propose to close this issue. I don't see this feature important while setup scripts itself are not ironed-out to trouble-free setup on Linux distributions. It is my view.
Downloading 30MB several times during manual install on one system is not issue for me. If Nix install deploys on scale, that it needs to be automatic (that script caching is not needed), prebuild in image/cached/setup must meet requirements of Nix installation.

And I try to contribute some pull requests to close couple of that issues.

**Comment by wmertens** [comment]
@Anton-Latukha The oneliner has to download the script, but the script could use the cached tarball. Between invocations, $TMPDIR is the same, but it's different per login. So the script could just check if the tarball exists in $TMPDIR/tarball.tgz with the correct size *and* hash, before attempting download.

I'm having a hard time imagining a doctored tgz file that has the same hash and size as the correct tgz file and also unpacks without problems.

Downloading even 30MB can be a lot. No reason to do it if we can just check for the cached file.

**Comment by Anton-Latukha** [comment]
@wmertens 
Current oneliner runs:
```shell
tmpDir="$(mktemp -d -t nix-binary-tarball-unpack.XXXXXXXXXX || \
          oops "Can\'t create temporary directory for downloading the Nix binary tarball")"
```
It is different for every invocation.
Maybe you talk about older version.

**Comment by wmertens** [comment]
My point is that it should be caching, not that it is doing it now…

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
